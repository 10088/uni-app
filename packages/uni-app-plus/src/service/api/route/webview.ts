import { formatLog } from '@dcloudio/uni-shared'
import { navigateFinish } from './utils'

export function closeWebview(
  webview: PlusWebviewWebviewObject,
  animationType: string,
  animationDuration: number
) {
  webview[(webview as any).__preload__ ? 'hide' : 'close'](
    animationType as any,
    animationDuration
  )
}

export function showWebview(
  webview: PlusWebviewWebviewObject,
  animationType: string,
  animationDuration: number,
  showCallback: Function,
  delay?: number
) {
  if (typeof delay === 'undefined') {
    delay = (webview as any).nvue ? 0 : 100
  }

  if (__DEV__) {
    console.log(formatLog('showWebview', 'delay', delay))
  }
  const execShowCallback = function () {
    if (execShowCallback._called) {
      if (__DEV__) {
        console.log(formatLog('execShowCallback', 'prevent'))
      }
      return
    }
    execShowCallback._called = true
    showCallback && showCallback()
    navigateFinish()
  }
  execShowCallback._called = false
  setTimeout(() => {
    const timer = setTimeout(() => {
      if (__DEV__) {
        console.log(formatLog('showWebview', 'callback', 'timer'))
      }
      execShowCallback()
    }, animationDuration + 150)

    webview.show(animationType as any, animationDuration, () => {
      if (__DEV__) {
        console.log(formatLog('showWebview', 'callback'))
      }
      if (!execShowCallback._called) {
        clearTimeout(timer)
      }
      execShowCallback()
    })
  }, delay)
}
