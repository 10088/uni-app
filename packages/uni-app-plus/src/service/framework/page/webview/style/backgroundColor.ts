import { isColor } from './utils'

export function initBackgroundColor(
  webviewStyle: PlusWebviewWebviewStyles,
  routeMeta: UniApp.PageRouteMeta
) {
  const { backgroundColor } = routeMeta
  if (!backgroundColor) {
    return
  }
  if (!isColor(backgroundColor)) {
    return
  }
  if (!webviewStyle.background) {
    webviewStyle.background = backgroundColor
  }
  if (!webviewStyle.backgroundColorTop) {
    webviewStyle.backgroundColorTop = backgroundColor
  }
}
