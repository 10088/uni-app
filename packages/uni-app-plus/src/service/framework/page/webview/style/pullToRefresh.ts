import { normalizePullToRefreshRpx } from '@dcloudio/uni-core'
import { extend } from '@vue/shared'

export function initPullToRefresh(
  webviewStyle: PlusWebviewWebviewStyles,
  routeMeta: UniApp.PageRouteMeta
) {
  if (!routeMeta.enablePullDownRefresh) {
    return
  }
  webviewStyle.pullToRefresh = normalizePullToRefreshRpx(
    extend(
      {},
      plus.os.name === 'Android'
        ? defaultAndroidPullToRefresh
        : defaultPullToRefresh,
      routeMeta.pullToRefresh
    )
  ) as unknown as PlusWebviewWebviewPullToRefreshStyles
}

const defaultAndroidPullToRefresh = { support: true, style: 'circle' }
const defaultPullToRefresh = {
  support: true,
  style: 'default',
  height: '50px',
  range: '200px',
  contentdown: {
    caption: '',
  },
  contentover: {
    caption: '',
  },
  contentrefresh: {
    caption: '',
  },
}
