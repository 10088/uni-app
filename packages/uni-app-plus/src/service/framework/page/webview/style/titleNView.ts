import { isArray } from '@vue/shared'
import { BACKGROUND_COLOR } from '@dcloudio/uni-shared'
import { isColor } from './utils'
import { invokeHook } from '@dcloudio/uni-core'
export function initTitleNView(
  webviewStyle: PlusWebviewWebviewStyles,
  routeMeta: UniApp.PageRouteMeta
) {
  const { navigationBar } = routeMeta
  if (navigationBar.style === 'custom') {
    return false
  }
  let autoBackButton = true
  if (routeMeta.isQuit) {
    autoBackButton = false
  }
  const titleNView: PlusWebviewWebviewTitleNViewStyles = {
    autoBackButton,
  }
  Object.keys(navigationBar).forEach((name) => {
    const value = navigationBar[name as keyof UniApp.PageNavigationBar]
    if (name === 'backgroundColor') {
      titleNView.backgroundColor = isColor(value as string)
        ? (value as string)
        : BACKGROUND_COLOR
    } else if (name === 'titleImage' && value) {
      titleNView.tags = createTitleImageTags(value as string)
    } else if (name === 'buttons' && isArray(value)) {
      titleNView.buttons = (value as UniApp.PageNavigationBar['buttons'])!.map(
        (button, index) => {
          ;(button as any).onclick = createTitleNViewBtnClick(index)
          return button
        }
      )
    }
  })

  webviewStyle.titleNView = titleNView
}

function createTitleImageTags(titleImage: string) {
  return [
    {
      tag: 'img',
      src: titleImage,
      position: {
        left: 'auto',
        top: 'auto',
        width: 'auto',
        height: '26px',
      },
    },
  ]
}

function createTitleNViewBtnClick(index: number) {
  return function onClick(btn: UniApp.PageNavigationBarButton) {
    ;(btn as any).index = index
    invokeHook('onNavigationBarButtonTap', btn)
  }
}
