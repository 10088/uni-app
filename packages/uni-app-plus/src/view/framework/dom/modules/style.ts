import { rpx2px } from '@dcloudio/uni-core'
import { camelize, capitalize, hyphenate, isArray, isString } from '@vue/shared'

export function patchStyle(el: Element, value: string | Record<string, any>) {
  const style = (el as HTMLElement).style
  if (isString(value)) {
    if (value === '') {
      el.removeAttribute('style')
    } else {
      // TODO display
      style.cssText = rpx2px(value, true)
    }
  } else {
    for (const key in value) {
      setStyle(style, key, value[key])
    }
  }
}

const importantRE = /\s*!important$/

function setStyle(
  style: CSSStyleDeclaration,
  name: string,
  val: string | string[]
) {
  if (isArray(val)) {
    val.forEach((v) => setStyle(style, name, v))
  } else {
    val = rpx2px(val, true)
    if (name.startsWith('--')) {
      // custom property definition
      style.setProperty(name, val)
    } else {
      const prefixed = autoPrefix(style, name)
      if (importantRE.test(val)) {
        // !important
        style.setProperty(
          hyphenate(prefixed),
          val.replace(importantRE, ''),
          'important'
        )
      } else {
        style[prefixed as any] = val
      }
    }
  }
}

// 移动端，仅处理 Webkit
const prefixes = ['Webkit' /*, 'Moz', 'ms'*/]
const prefixCache: Record<string, string> = {}

function autoPrefix(style: CSSStyleDeclaration, rawName: string): string {
  const cached = prefixCache[rawName]
  if (cached) {
    return cached
  }
  let name = camelize(rawName)
  if (name !== 'filter' && name in style) {
    return (prefixCache[rawName] = name)
  }
  name = capitalize(name)
  for (let i = 0; i < prefixes.length; i++) {
    const prefixed = prefixes[i] + name
    if (prefixed in style) {
      return (prefixCache[rawName] = prefixed)
    }
  }
  return rawName
}
