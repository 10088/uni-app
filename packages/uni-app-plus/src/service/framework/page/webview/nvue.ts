import { extend } from '@vue/shared'

import { CreateWebviewOptions } from '.'
import { parseWebviewStyle } from './style'
import { genWebviewId, initUniPageUrl } from './utils'

export function createNVueWebview({
  path,
  query,
  routeOptions,
  webviewStyle,
}: CreateWebviewOptions) {
  const curWebviewId = genWebviewId()
  const curWebviewStyle = parseWebviewStyle(curWebviewId, path, routeOptions)
  ;(curWebviewStyle as any).uniPageUrl = initUniPageUrl(path, query)
  if (__DEV__) {
    console.log('[uni-app] createWebview', curWebviewId, path, curWebviewStyle)
  }
  // android 需要使用
  ;(curWebviewStyle as any).isTab = !!routeOptions.meta.isTabBar
  return plus.webview.create(
    '',
    String(curWebviewId),
    curWebviewStyle,
    extend(
      {
        nvue: true,
      },
      webviewStyle
    )
  )
}
