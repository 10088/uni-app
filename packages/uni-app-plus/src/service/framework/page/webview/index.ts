import { createNVueWebview } from './nvue'
import { getPreloadWebview, getWebviewId } from './utils'

export interface CreateWebviewOptions {
  path: string
  query: Record<string, string>
  routeOptions: UniApp.UniRoute
  webviewStyle?: Record<string, any>
}

export function createWebview(options: CreateWebviewOptions) {
  if (options.routeOptions.meta.isNVue) {
    return createNVueWebview(options)
  }
  if (getWebviewId() === 2) {
    // 如果首页非 nvue，则直接返回 Launch Webview
    return plus.webview.getLaunchWebview()
  }
  return getPreloadWebview()
}
