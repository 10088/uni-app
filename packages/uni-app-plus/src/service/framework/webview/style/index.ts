import { initNVue } from './nvue'
import { initBackgroundColor } from './backgroundColor'
import { initPopGesture } from './popGesture'
import { initPullToRefresh } from './pullToRefresh'
import { initTitleNView } from './titleNView'
import { DebugRefresh, InitUniPageUrl } from '../utils'

export function parseWebviewStyle(
  path: string,
  routeMeta: UniApp.PageRouteMeta
): PlusWebviewWebviewStyles & {
  uniPageUrl?: InitUniPageUrl
  debugRefresh?: DebugRefresh
  isTab?: boolean
} {
  const webviewStyle: PlusWebviewWebviewStyles = {
    bounce: 'vertical',
  }

  Object.keys(routeMeta).forEach((name) => {
    if (WEBVIEW_STYLE_BLACKLIST.indexOf(name) === -1) {
      webviewStyle[name as keyof PlusWebviewWebviewStyles] =
        routeMeta[name as keyof UniApp.PageRouteMeta]
    }
  })
  initNVue(webviewStyle, routeMeta, path)
  initPopGesture(webviewStyle, routeMeta)
  initBackgroundColor(webviewStyle, routeMeta)
  initTitleNView(webviewStyle, routeMeta)
  initPullToRefresh(webviewStyle, routeMeta)

  return webviewStyle
}

const WEBVIEW_STYLE_BLACKLIST = [
  'id',
  'route',
  'isNVue',
  'isQuit',
  'isEntry',
  'isTabBar',
  'tabBarIndex',
  'windowTop',
  'topWindow',
  'leftWindow',
  'rightWindow',
  'maxWidth',
  'usingComponents',
  'disableScroll',
  'enablePullDownRefresh',
  'navigationBar',
  'pullToRefresh',
  'onReachBottomDistance',
  'pageOrientation',
  'backgroundColor',
]
