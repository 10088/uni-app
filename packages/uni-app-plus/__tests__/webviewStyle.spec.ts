import { parseWebviewStyle } from '../src/service/framework/webview/style'
function initDefaultUniConfig() {
  return JSON.parse(
    JSON.stringify({
      globalStyle: {
        navigationBar: {},
      },
    })
  )
}
beforeAll(() => {
  global.__uniConfig = initDefaultUniConfig()
})
const defaultPath = '/pages/index/index'
const defaultRoute = 'pages/index/index'
describe('webviewStyle', () => {
  test('basic', () => {
    expect(
      parseWebviewStyle(1, defaultPath, {
        path: defaultPath,
        meta: {
          route: defaultRoute,
          navigationBar: {},
        },
      })
    ).toMatchSnapshot()
  })
})
