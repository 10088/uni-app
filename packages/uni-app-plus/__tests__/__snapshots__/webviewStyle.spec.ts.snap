// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webviewStyle basic 1`] = `
Object {
  "bounce": "vertical",
  "titleNView": Object {
    "autoBackButton": true,
  },
}
`;
